File: 프로그래머스/2/1844. 게임 맵 최단거리/게임 맵 최단거리.py
# BFS 최단거리 2차원 배열
from collections import deque

def solution(maps):
    dx = [1, 0, -1, 0]
    dy = [0, 1, 0, -1]
    n, m = len(maps), len(maps[0])
    
    visited = set([(0, 0)])
    queue = deque([(0, 0, 1)])
    
    while queue:
        x, y, dist = queue.popleft()

        if (x == n-1) and (y == m-1):
            return dist

        for i in range(4):
            nx = x + dx[i]
            ny = y + dy[i]

            if (0 <= nx < n and 0 <= ny < m) and (maps[nx][ny]!=0):
                if (nx, ny) not in visited:
                    visited.add((nx, ny))
                    queue.append((nx, ny, dist + 1))
    return -1

File: 프로그래머스/2/1844. 게임 맵 최단거리/게임 맵 최단거리.java
import java.util.*;
class Solution {
    public int solution(int[][] maps) {
        
        int n = maps.length;
        int m = maps[0].length;
        
        int[] loc = {0, 0};
        int[] fin = {maps.length-1, maps[0].length-1};
        
        Queue<int[]> qq = new ArrayDeque<>();
        
        boolean[][] visited = new boolean[n][m];
        
        int[] dx = {1,0,-1,0};
        int[] dy = {0,1,0,-1};
        
        qq.offer(new int[] {0, 0, 1});
        visited[0][0] = true;
        
        while (!qq.isEmpty()) {
            int[] a = qq.poll();
            int x = a[0];
            int y = a[1];
            int cnt = a[2];
            // System.out.println(x + " " + y);
            
            if (x == maps.length -1 && y == maps[0].length -1 ) {
                return cnt;
            }
            
            for (int i=0; i<4; i++) {
                int newx = x + dx[i];
                int newy = y + dy[i];
                // System.out.println(": " + newx + " " + newy);
                if (0 <= newx && newx < maps.length && 0 <= newy && newy < maps[0].length && maps[newx][newy] == 1 && visited[newx][newy] != true) {
                    qq.offer(new int[] {newx, newy, cnt +1});
                    visited[newx][newy] = true;
                }
            }
        }
        
        return -1;

    }
}